### PR TITLE
Rescore SSMs using semi-supervised learning

### PR DESCRIPTION
**Pull request by issararab**
Rescore SSMs using semi-supervised learning with mokapot.

Contributions:

- Integration of mokapot to rescore SSMs.
- Various spectrum similarity measures to be used as features for mokapot machine learning.

### CODE REVIEW

**Comment by bittremieux** [comment]
@wfondrie We're making progress with the mokapot integration. :) What is your experience with the need to perform hyperparameter tuning of classifiers (incl model choice) inside of mokapot? Our initial tests indicated a significant performance difference between a linear SVM and RF.

**Comment by wfondrie** [comment]
It depends on the particular search engine and features, but I also tend to see performance improvements with other types of tree-based models (XGBoost). That's not always the case though - in particularly, the linear SVM works pretty well for a standard DB search with Tide. 

As for tuning hyperparameters, sklearn's `GridSearchCV` and `RandomizedSearchCV` should be fully compatible for use as the `estimator` in `mokapot.Model()`. As for choosing between multiple types of models, you'll have to write something more custom. As long as it acts like sklearn's `GridSearchCV`, it should work though!

**Comment by bittremieux** [comment]
Ok, that matches our evaluations so far. I was mainly wondering whether there are specific best practices or not.

I think we'll stick to a random forest rather than XGBoost, because RF is less sensitive to hyperparameters. We'll have the option to switch between RF (default) and a Percolator-like SVM. RF seemed to do especially better for the open search results, which intuitively makes sense because a non-linear classifier might be able to better capture the different groups.

Do you maybe also have some suggestions in terms of features? We have a decent set already, but additional information-rich features is always better of course.

**Comment by wfondrie** [comment]
> Do you maybe also have some suggestions in terms of features? 
Not really other than the usual suspects---make sure that charge is included (preferably one-hot encoded).  

The real key is just to make sure we don't include features that compromise the integrity of the FDR estimates. These are typically features that describe peptide properties in the database search setting, like amino-acid or k-mer frequencies, but I'm not sure what analogous feature would be for the library search. 

I would recommend testing your final feature set empirically with an entrapment experiment to verify that we're not getting liberal FDR estimates due to such a feature (or more sneaky, a combination of features). I've never done this for a library search, but my inclination would be to append another larger spectral library to the target library and use that as the entrapment database.

File: src/ann_solo/spectrum_similarity.py [hunk 1]
```diff
@@ -0,0 +1,878 @@
...
+        ssm.query_spectrum.intensity[ssm.peak_matches[:, 0]],
+        ssm.library_spectrum.intensity[ssm.peak_matches[:, 1]],
+    )
+
+
+def improved_sim(ssm: spectrum.SpectrumSpectrumMatch) -> float:
+    """
+    Get the Improved Similarity Index of peak matches between two spectra.
+
+    TODO: reference needed.
```
Comment (bittremieux) [inline src/ann_solo/spectrum_similarity.py:722]: @issararab Do you have a (mass spec) reference where this "improved similarity" comes from?

File: src/ann_solo/utils.py [hunk 2]
```diff
@@ -1,104 +1,338 @@
...
+                scaler,
+                train_fdr=fdr,
+            )
         else:
-            break
+            raise ValueError(
+                "Unknown semi-supervised machine learning model given"
+            )
+        # Train the mokapot model and combine the SSMs for all groups.
+        confidences, _ = mokapot.brew(dataset, clf, fdr, folds=5)
```
Comment (bittremieux) [inline src/ann_solo/utils.py:133]: Let's stick with the default 3-fold cross-validation as also used by Percolator. This is accepted practice by the community, and I don't see a special reason that 5-fold CV would be better.

File: src/ann_solo/utils.py [hunk 3]
```diff
@@ -1,104 +1,338 @@
...
+        features["scribe_fragment_acc"].append(sim.scribe_fragment_acc(ssm))
+        features["manhattan"].append(sim.manhattan(ssm))
+        features["euclidean"].append(sim.euclidean(ssm))
+        features["chebyshev"].append(sim.chebyshev(ssm))
+        features["pearsonr"].append(sim.pearsonr(ssm))
+        features["braycurtis"].append(sim.braycurtis(ssm))
+        features["canberra"].append(sim.canberra(ssm))
+        features["ruzicka"].append(sim.ruzicka(ssm))
+        features["is_target"].append(not ssm.is_decoy)
+        features["group"].append("G")
```
Comment (bittremieux) [inline src/ann_solo/utils.py:337]: I don't understand this change. This would assign everything to the same group, rather than the groups per modification that we want.

File: src/ann_solo/utils.py [hunk 4]
```diff
@@ -1,104 +1,338 @@
...
+    )
+    # Assign the scores to the SSM objects.
+    for i, score, q in zip(
+        ssm_scores["index"],
+        ssm_scores["mokapot score"],
+        ssm_scores["mokapot q-value"],
+    ):
+        ssms[i].search_engine_score = score
+        ssms[i].q = q
+        yield ssms[i]
```
Comment (bittremieux) [inline src/ann_solo/utils.py:151]: There's not really an advantage of using an iterator here, because we've already consumed all SSMs previously anyway. I suggest reverting to the `return ...`. Alternatively, the docstring and function type need to be updated accordingly.

File: src/ann_solo/utils.py [hunk 5]
```diff
@@ -1,104 +1,338 @@
...
         else:
-            break
+            raise ValueError(
+                "Unknown semi-supervised machine learning model given"
+            )
+        # Train the mokapot model and combine the SSMs for all groups.
+        confidences, _ = mokapot.brew(dataset, clf, fdr, folds=5)
+    ssm_scores = pd.concat(
+        [
+            confidences.group_confidence_estimates[group].psms[
```
Comment (bittremieux) [inline src/ann_solo/utils.py:136]: I would suggest making a small change how ANN-SoLo reports results to include all SSMs, not only those below the specified FDR threshold. That can be useful for downstream analyses. In that case, maybe doing the filtering at a higher level in the code will give a better design.

File: src/ann_solo/utils.py (hunk 3 above)
Comment (issararab) [reply src/ann_solo/utils.py:337]: In the code you changed previously, you assign the group from the class SpectrumSpectrumMatch which doesn't even exist as an attribute, while executing the whole code was getting an error here. Second, this line of code is kinda useless as we call _get_ssm_groups() in the next statement (Line 74) which assigns new groups for this column. This line (337) can even be omitted.

Comment (bittremieux) [reply src/ann_solo/utils.py:337]: Oh, that's right. Must have been some incomplete changes on my part. Yes, better to just remove this line here then, rather than have it be confusing.

File: src/ann_solo/utils.py (hunk 4 above)
Comment (issararab) [reply src/ann_solo/utils.py:151]: The return statement retruns the whole list of PSM matches found by ANN-SoLo after assigning the new score and the q-value for all positive targets. What we need is just the true target entries matching our FDR to be returned as our best matches (same logic you had previously in the old ANN-SoLo version). Otherwise, you always get the full list as the input query.

This statement: confidences.group_confidence_estimates[group].psms , returns all True targets with their corresponding score and q-value. So I changed that to filter for entries above the FDR threshold in the concatenation then return those matches later on in the above loop.

Comment (bittremieux) [reply src/ann_solo/utils.py:151]: Indeed, I think it could be useful to separate rescoring from FDR filtering. So first return the full list from this function, and then in the calling function that implements the cascade strategy do the filtering and submitting the remaining spectra to the second round. That way we can uncouple unrelated functionality between (re)scoring and FDR filtering.

So your changes are correct (thx!), I'd just put them in a different location in the code.

File: src/ann_solo/utils.py (hunk 5 above)
Comment (issararab) [reply src/ann_solo/utils.py:136]: We can discuss it tomorrow whether we want to include all True Targets at the end or just the matches passing the FDR threshold. It will also be hard to benchmark now that we have all results of other tools. I believe they always return the best matching true targets above the threshold at the end. But we can discuss a better solution.

File: src/ann_solo/utils.py (hunk 4 above)
Comment (issararab) [reply src/ann_solo/utils.py:151]: Would also work. Was just following the same logic we had in the current version of ANN-SoLo to have correct output as I was running the whole software.

Comment (issararab) [reply src/ann_solo/utils.py:151]: Wanted to get some results on our large test set, but CalCua is down for 4 days this week :/

Comment (bittremieux) [reply src/ann_solo/utils.py:151]: Yes, makes sense. This indeed follows how ANN-SoLo was doing it originally, but I've since realized that that's not always the best solution. But this is not documented, so impossible to know that in advance of course.

File: src/ann_solo/utils.py (hunk 2 above)
Comment (issararab) [reply src/ann_solo/utils.py:133]: This is a hyperparameter that I want to investigate once the HPC is available again. If I recall well, the initial paper of Percolator mentioned a 10-fold CV and stops earlier if it converges already before. The default in Mokpot is 3-folds. We can try 3 / 5 / 10 in the 3 query files I am currently using.

Comment (bittremieux) [reply src/ann_solo/utils.py:133]: The number of CV folds isn't usually a hyperparameter that necessarily needs to be tuned, instead it depends on your computational budget. There are a few papers that explain the Percolator CV procedure in more detail (e.g. [Granholm et al.](http://www.biomedcentral.com/1471-2105/13/S16/S3)). It usually doesn't really make a difference for performance, so I think here we can safely stick to the default.

Comment (issararab) [reply src/ann_solo/utils.py:133]: Thanks Wout for sharing the reference with me. Interesting paper. In that case, we stick to the default k-fold CV used by Mokapot.

File: src/ann_solo/spectral_library.py [hunk 6]
```diff
@@ -103,7 +103,8 @@ def __init__(self, filename: str) -> None:
                            self._library_reader.spec_info['charge'].items()
                            if len(charge_info['id']) >= config.num_list]
             for charge in sorted(ann_charges):
-                self._ann_filenames[charge] = f'{base_filename}_{charge}.idxann'
+                self._ann_filenames[charge] = \
+                    f'{base_filename}_{charge}.idxann'
```
Comment (wfondrie) [inline src/ann_solo/spectral_library.py:107]: **nitpick:** I typically prefer using parenthesis instead of backslash for line breaks.
```python
self._ann_filenames[charge] = (
    f'{base_filename}_{charge}.idxann'
)
```

File: src/ann_solo/spectrum.py [hunk 7]
```diff
@@ -17,8 +18,13 @@ def _check_spectrum_valid(spectrum_mz: np.ndarray, min_peaks: int,
 
     Parameters
     ----------
-    spectrum : np.ndarray
+    spectrum_mz : np.ndarray
         M/z peaks of the sspectrum whose quality is checked.
```
Comment (wfondrie) [inline src/ann_solo/spectrum.py:22]: **typo:** Not part of this PR, but `sspectrum` -> `spectrum`.

File: src/ann_solo/spectrum_similarity.py [hunk 8]
```diff
@@ -0,0 +1,741 @@
...
+
+from ann_solo import spectrum
+
+
+class SpectrumSimilarityCalculator:
+    def __init__(
+        self, ssm: spectrum.SpectrumSpectrumMatch, top: Optional[int] = None
+    ):
+        """
+        Instantiate the `SpectrumSimilarityFactory` to compute various spectrum
```
Comment (wfondrie) [inline src/ann_solo/spectrum_similarity.py:18]: **question:** Did you mean `SpectrumSimilarityCalculator` here? If not, I wouldn't really call this a true factory pattern, although that's pretty nitpicky. Feel free to ignore!

File: src/ann_solo/spectrum_similarity.py [hunk 9]
```diff
@@ -0,0 +1,741 @@
...
+                            )
+                            / scipy.special.comb(n_peak_bins, n_library_peaks)
+                            for i in range(
+                                n_matched_peaks + 1, n_library_peaks + 1
+                            )
+                        ]
+                    )
+                ),
+                100.0,
+            )
```
Comment (wfondrie) [inline src/ann_solo/spectrum_similarity.py:317]: **suggestion:** Consider reformatting this, because it's difficult to follow.

File: src/ann_solo/utils.py [hunk 10]
```diff
@@ -1,104 +1,457 @@
...
+                            {0: 1, 1: 10},
+                            {0: 10, 1: 0.1},
+                            {0: 10, 1: 1},
+                        ],
+                    },
+                    refit=False,
+                    cv=3,
+                    n_jobs=-1,
+                ),
+                scaler,
```
Comment (wfondrie) [inline src/ann_solo/utils.py:176]: **praise:** This is super cool! I hadn't actually thought about using a pipeline here.

File: src/ann_solo/utils.py [hunk 11]
```diff
@@ -1,104 +1,457 @@
...
+        ssms[i].search_engine_score = score
+        ssms[i].q = q
+    return ssms
 
 
-def filter_group_fdr(ssms: Iterator[SpectrumSpectrumMatch], fdr: float = 0.01,
-                     tol_mass: float = None, tol_mode: str = None,
-                     min_group_size: int = None)\
-        -> Iterator[SpectrumSpectrumMatch]:
+def _get_ssm_groups(
```
Comment (wfondrie) [inline src/ann_solo/utils.py:204]: **question:** I've thought about adding this functionality in mokapot, but I'm not sure if its worth it. What do you think?

File: src/ann_solo/utils.py [hunk 12]
```diff
@@ -1,104 +1,457 @@
...
+        features["chebyshev"].append(sim_calc.chebyshev())
+        features["pearsonr"].append(sim_calc.pearsonr())
+        features["pearsonr_top5"].append(sim_calc_top.pearsonr())
+        features["spearmanr"].append(sim_calc.spearmanr())
+        features["spearmanr_top5"].append(sim_calc_top.spearmanr())
+        features["braycurtis"].append(sim_calc.braycurtis())
+        features["canberra"].append(sim_calc.canberra())
+        features["ruzicka"].append(sim_calc.ruzicka())
+        features["is_target"].append(not ssm.is_decoy)
+    return features
```
Comment (wfondrie) [inline src/ann_solo/utils.py:457]: **question:** Is performance not a problem here?

**Review by wfondrie — COMMENTED** [review]
This looks pretty good to me. My main suggestion is a code formatting one: the spectrum similarity calculations could be better formatted for readability. 

Aside from that, I think it was mostly questions and nitpick comments.

File: src/ann_solo/spectral_library.py (hunk 6 above)
Comment (bittremieux) [reply src/ann_solo/spectral_library.py:107]: Agree. Although eventually I need to use black on the entire project to get a consistent and enforced code style everywhere. I didn't want to do it yet while we're working on this PR to make it easier to see what's changed. We'll use black after this functionality is completed.

File: src/ann_solo/spectrum_similarity.py (hunk 8 above)
Comment (bittremieux) [reply src/ann_solo/spectrum_similarity.py:18]: Thanks. This was lingering from a previous naming. Because it's indeed not really a factory pattern, I switched to `SpectrumSimilarityCalculator` from the initial `SpectrumSimilarityFactory`. The former is a better name, but the refactoring didn't catch the docstring apparently.

File: src/ann_solo/utils.py (hunk 11 above)
Comment (bittremieux) [reply src/ann_solo/utils.py:204]: To my knowledge there's no "default" way of grouping PSMs that everyone does, so there's no generally accepted way that mokapot would do it. Some custom implementation could be useful, but I don't consider it core functionality that's absolutely necessary.

Coincidentally this is very similar to how the PTM-Shepherd paper describes that they do it though, even though it was developed independently for another project without having read that paper. It would be useful to further validate this part of my code with unit tests on a representative example though to be sure it fully works as expected.

File: src/ann_solo/utils.py (hunk 12 above)
Comment (bittremieux) [reply src/ann_solo/utils.py:457]: Agree. This can and should be optimized, i.e. by multithreading and by using Numba. My main focus was getting the full functionality implemented first, but we can probably think about performance now as well. Multithreading should normally be a trivial change.

**Review by bittremieux — APPROVED** [review]
